File: .claude/agents/wiki-gcp-agent/scripts/gcp_query.py
#!/usr/bin/env python3
from __future__ import annotations

"""GCP data fetcher for the wiki-gcp-agent.

Library usage:
    from gcp_query import fetch
    result = fetch("list_services", {"project_id": "acme-prod-123"})

CLI usage:
    python gcp_query.py --action list_services --params '{"project_id": "acme-prod-123"}'
    python gcp_query.py --action describe_db --params '{"project_id": "acme-prod-123", "instance_id": "main-pg"}'
    python gcp_query.py --action list_topics --params '{"project_id": "acme-prod-123"}'
    python gcp_query.py --action query_logs --params '{"project_id": "acme-prod-123", "filter": "severity>=ERROR"}'
"""

import argparse
import json
import re
import sys
from typing import Any

VALID_ACTIONS = {"list_services", "describe_db", "list_topics", "query_logs"}

MAX_RESULTS_DEFAULT = 100
MAX_RESULTS_CAP = 1000
MAX_LOG_HOURS = 168  # 7 days

PROJECT_ID_PATTERN = re.compile(r"^[a-z][a-z0-9-]{4,28}[a-z0-9]$")


def _validate_project_id(project_id: str) -> str:
    """Validate GCP project ID format."""
    if not PROJECT_ID_PATTERN.match(project_id):
        raise ValueError(
            f"Invalid project_id '{project_id}' — must be 6-30 lowercase letters, digits, hyphens"
        )
    return project_id


def _validate_list_services(params: dict[str, Any]) -> dict[str, Any]:
    """Validate params for list_services action."""
    project_id = _validate_project_id(params.get("project_id", ""))
    return {"project_id": project_id}


def _validate_describe_db(params: dict[str, Any]) -> dict[str, Any]:
    """Validate params for describe_db action."""
    project_id = _validate_project_id(params.get("project_id", ""))
    instance_id = params.get("instance_id", "")
    if not instance_id or not isinstance(instance_id, str):
        raise ValueError("describe_db requires a non-empty 'instance_id' string")
    database = params.get("database", "")
    return {"project_id": project_id, "instance_id": instance_id, "database": database}


def _validate_list_topics(params: dict[str, Any]) -> dict[str, Any]:
    """Validate params for list_topics action."""
    project_id = _validate_project_id(params.get("project_id", ""))
    return {"project_id": project_id}


def _validate_query_logs(params: dict[str, Any]) -> dict[str, Any]:
    """Validate params for query_logs action."""
    project_id = _validate_project_id(params.get("project_id", ""))
    log_filter = params.get("filter", "")
    if not log_filter or not isinstance(log_filter, str):
        raise ValueError("query_logs requires a non-empty 'filter' string")
    # Basic injection prevention: reject semicolons and quotes
    if ";" in log_filter or "'" in log_filter or '"' in log_filter:
        raise ValueError("Filter contains forbidden characters — no semicolons or quotes allowed")
    hours = min(int(params.get("hours", 24)), MAX_LOG_HOURS)
    max_results = min(
        int(params.get("max_results", MAX_RESULTS_DEFAULT)),
        MAX_RESULTS_CAP,
    )
    return {
        "project_id": project_id,
        "filter": log_filter.strip(),
        "hours": hours,
        "max_results": max_results,
    }


_VALIDATORS = {
    "list_services": _validate_list_services,
    "describe_db": _validate_describe_db,
    "list_topics": _validate_list_topics,
    "query_logs": _validate_query_logs,
}


def fetch(action: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
    """Fetch data from GCP.

    Args:
        action: One of list_services, describe_db, list_topics, query_logs.
        params: Action-specific parameters.

    Returns:
        Structured result dict with status, action, and data fields.
    """
    if action not in VALID_ACTIONS:
        return {
            "status": "error",
            "error_code": "VALIDATION_ERROR",
            "message": f"Unknown action '{action}' — expected one of {sorted(VALID_ACTIONS)}",
        }

    params = params or {}

    try:
        validated = _VALIDATORS[action](params)
    except (ValueError, TypeError) as exc:
        return {
            "status": "error",
            "error_code": "VALIDATION_ERROR",
            "message": str(exc),
        }

    try:
        handler = _HANDLERS[action]
        return handler(validated)
    except Exception as exc:  # noqa: BLE001
        return {
            "status": "error",
            "error_code": "CONNECTION_ERROR",
            "message": f"GCP API call failed: {exc}",
        }


def _fetch_list_services(validated: dict[str, Any]) -> dict[str, Any]:
    """List GCP services in project."""
    # PLACEHOLDER: actual GCP API call
    # from google.cloud import run_v2
    # client = run_v2.ServicesClient()
    # services = client.list_services(parent=f"projects/{validated['project_id']}/locations/-")
    return {
        "status": "success",
        "action": "list_services",
        "data": {
            "project_id": validated["project_id"],
            "services": [],
            "service_count": 0,
        },
    }


def _fetch_describe_db(validated: dict[str, Any]) -> dict[str, Any]:
    """Describe a Cloud SQL instance."""
    # PLACEHOLDER: actual GCP API call
    # from google.cloud import sqladmin_v1
    # client = sqladmin_v1.SqlInstancesServiceClient()
    # instance = client.get(project=validated["project_id"], instance=validated["instance_id"])
    return {
        "status": "success",
        "action": "describe_db",
        "data": {
            "project_id": validated["project_id"],
            "instance_id": validated["instance_id"],
            "database": validated.get("database", ""),
            "engine": "",
            "version": "",
            "tier": "",
            "region": "",
            "state": "",
            "tables": [],
        },
    }


def _fetch_list_topics(validated: dict[str, Any]) -> dict[str, Any]:
    """List Pub/Sub topics in project."""
    # PLACEHOLDER: actual GCP API call
    # from google.cloud import pubsub_v1
    # publisher = pubsub_v1.PublisherClient()
    # topics = publisher.list_topics(project=f"projects/{validated['project_id']}")
    return {
        "status": "success",
        "action": "list_topics",
        "data": {
            "project_id": validated["project_id"],
            "topics": [],
            "topic_count": 0,
        },
    }


def _fetch_query_logs(validated: dict[str, Any]) -> dict[str, Any]:
    """Query Cloud Logging entries."""
    # PLACEHOLDER: actual GCP API call
    # from google.cloud import logging_v2
    # client = logging_v2.Client(project=validated["project_id"])
    # entries = client.list_entries(filter_=validated["filter"], max_results=validated["max_results"])
    return {
        "status": "success",
        "action": "query_logs",
        "data": {
            "project_id": validated["project_id"],
            "filter": validated["filter"],
            "hours": validated["hours"],
            "entries": [],
            "entry_count": 0,
        },
        "truncated": False,
    }


_HANDLERS = {
    "list_services": _fetch_list_services,
    "describe_db": _fetch_describe_db,
    "list_topics": _fetch_list_topics,
    "query_logs": _fetch_query_logs,
}


def main() -> None:
    """CLI entry point."""
    parser = argparse.ArgumentParser(description="Query GCP resources")
    parser.add_argument(
        "--action",
        required=True,
        choices=sorted(VALID_ACTIONS),
        help="Action to perform",
    )
    parser.add_argument(
        "--params",
        default="{}",
        help="JSON string of action parameters",
    )
    args = parser.parse_args()

    try:
        params = json.loads(args.params)
    except json.JSONDecodeError as exc:
        result = {
            "status": "error",
            "error_code": "VALIDATION_ERROR",
            "message": f"Invalid JSON in --params: {exc}",
        }
        print(json.dumps(result, indent=2))
        sys.exit(1)

    result = fetch(args.action, params)
    print(json.dumps(result, indent=2))

    if result["status"] != "success":
        sys.exit(1)


if __name__ == "__main__":
    main()

File: .claude/agents/wiki-gcp-agent/scripts/gcp_query.ts
#!/usr/bin/env node
/**
 * GCP data fetcher for the wiki-gcp-agent.
 *
 * Library usage:
 *     import { fetch } from "./gcp_query.js";
 *     const result = fetch("list_services", { project_id: "acme-prod-123" });
 *
 * CLI usage:
 *     node gcp_query.js --action list_services --params '{"project_id": "acme-prod-123"}'
 *     node gcp_query.js --action describe_db --params '{"project_id": "acme-prod-123", "instance_id": "main-pg"}'
 *     node gcp_query.js --action list_topics --params '{"project_id": "acme-prod-123"}'
 *     node gcp_query.js --action query_logs --params '{"project_id": "acme-prod-123", "filter": "severity>=ERROR"}'
 *
 * This is a TypeScript port of gcp_query.py. The Python reference has
 * commented-out google-cloud-* calls (stubbed); this TS port matches that
 * stub so CLI JSON output is byte-identical.
 */
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { pythonJsonDumps } from "../../../skills/wiki/scripts/_json_py.js";

// ── Constants ───────────────────────────────────────────────────────

export const VALID_ACTIONS: ReadonlySet<string> = new Set([
  "list_services",
  "describe_db",
  "list_topics",
  "query_logs",
]);

const MAX_RESULTS_DEFAULT = 100;
const MAX_RESULTS_CAP = 1000;
const MAX_LOG_HOURS = 168; // 7 days

const PROJECT_ID_PATTERN = /^[a-z][a-z0-9-]{4,28}[a-z0-9]$/;

export type FetchResult = Record<string, unknown>;
type Params = Record<string, unknown>;

function toInt(value: unknown, fallback: number): number {
  if (typeof value === "number" && Number.isFinite(value)) {
    return Math.trunc(value);
  }
  if (typeof value === "string") {
    const n = parseInt(value, 10);
    if (Number.isFinite(n)) return n;
  }
  return fallback;
}

function validateProjectId(projectId: string): string {
  if (!PROJECT_ID_PATTERN.test(projectId)) {
    throw new Error(
      `Invalid project_id '${projectId}' — must be 6-30 lowercase letters, digits, hyphens`,
    );
  }
  return projectId;
}

interface ListServicesValidated {
  project_id: string;
}

interface DescribeDbValidated {
  project_id: string;
  instance_id: string;
  database: string;
}

interface ListTopicsValidated {
  project_id: string;
}

interface QueryLogsValidated {
  project_id: string;
  filter: string;
  hours: number;
  max_results: number;
}

function validateListServices(params: Params): ListServicesValidated {
  const raw = params["project_id"];
  const projectId = validateProjectId(typeof raw === "string" ? raw : "");
  return { project_id: projectId };
}

function validateDescribeDb(params: Params): DescribeDbValidated {
  const raw = params["project_id"];
  const projectId = validateProjectId(typeof raw === "string" ? raw : "");
  const instIdRaw = params["instance_id"];
  if (!instIdRaw || typeof instIdRaw !== "string") {
    throw new Error("describe_db requires a non-empty 'instance_id' string");
  }
  const dbRaw = params["database"] ?? "";
  const database = typeof dbRaw === "string" ? dbRaw : "";
  return { project_id: projectId, instance_id: instIdRaw, database };
}

function validateListTopics(params: Params): ListTopicsValidated {
  const raw = params["project_id"];
  const projectId = validateProjectId(typeof raw === "string" ? raw : "");
  return { project_id: projectId };
}

function validateQueryLogs(params: Params): QueryLogsValidated {
  const raw = params["project_id"];
  const projectId = validateProjectId(typeof raw === "string" ? raw : "");
  const filterRaw = params["filter"];
  if (!filterRaw || typeof filterRaw !== "string") {
    throw new Error("query_logs requires a non-empty 'filter' string");
  }
  if (
    filterRaw.includes(";") ||
    filterRaw.includes("'") ||
    filterRaw.includes('"')
  ) {
    throw new Error(
      "Filter contains forbidden characters — no semicolons or quotes allowed",
    );
  }
  const hours = Math.min(toInt(params["hours"], 24), MAX_LOG_HOURS);
  const maxResults = Math.min(
    toInt(params["max_results"], MAX_RESULTS_DEFAULT),
    MAX_RESULTS_CAP,
  );
  return {
    project_id: projectId,
    filter: filterRaw.trim(),
    hours,
    max_results: maxResults,
  };
}

function fetchListServices(validated: ListServicesValidated): FetchResult {
  return {
    status: "success",
    action: "list_services",
    data: {
      project_id: validated.project_id,
      services: [],
      service_count: 0,
    },
  };
}

function fetchDescribeDb(validated: DescribeDbValidated): FetchResult {
  return {
    status: "success",
    action: "describe_db",
    data: {
      project_id: validated.project_id,
      instance_id: validated.instance_id,
      database: validated.database,
      engine: "",
      version: "",
      tier: "",
      region: "",
      state: "",
      tables: [],
    },
  };
}

function fetchListTopics(validated: ListTopicsValidated): FetchResult {
  return {
    status: "success",
    action: "list_topics",
    data: {
      project_id: validated.project_id,
      topics: [],
      topic_count: 0,
    },
  };
}

function fetchQueryLogs(validated: QueryLogsValidated): FetchResult {
  return {
    status: "success",
    action: "query_logs",
    data: {
      project_id: validated.project_id,
      filter: validated.filter,
      hours: validated.hours,
      entries: [],
      entry_count: 0,
    },
    truncated: false,
  };
}

/**
 * Fetch data from GCP.
 */
export function fetch(
  action: string,
  params: Params | null = null,
): FetchResult {
  if (!VALID_ACTIONS.has(action)) {
    const sorted = [...VALID_ACTIONS].sort();
    return {
      status: "error",
      error_code: "VALIDATION_ERROR",
      message:
        `Unknown action '${action}' — expected one of ` +
        `[${sorted.map((s) => `'${s}'`).join(", ")}]`,
    };
  }

  const p: Params = params ?? {};

  let validated:
    | ListServicesValidated
    | DescribeDbValidated
    | ListTopicsValidated
    | QueryLogsValidated;
  try {
    switch (action) {
      case "list_services":
        validated = validateListServices(p);
        break;
      case "describe_db":
        validated = validateDescribeDb(p);
        break;
      case "list_topics":
        validated = validateListTopics(p);
        break;
      case "query_logs":
        validated = validateQueryLogs(p);
        break;
      default:
        throw new Error("unreachable");
    }
  } catch (exc) {
    return {
      status: "error",
      error_code: "VALIDATION_ERROR",
      message: (exc as Error).message,
    };
  }

  try {
    switch (action) {
      case "list_services":
        return fetchListServices(validated as ListServicesValidated);
      case "describe_db":
        return fetchDescribeDb(validated as DescribeDbValidated);
      case "list_topics":
        return fetchListTopics(validated as ListTopicsValidated);
      case "query_logs":
        return fetchQueryLogs(validated as QueryLogsValidated);
    }
  } catch (exc) {
    return {
      status: "error",
      error_code: "CONNECTION_ERROR",
      message: `GCP API call failed: ${(exc as Error).message}`,
    };
  }

  return {
    status: "error",
    error_code: "UNKNOWN",
    message: "Unexpected state",
  };
}

// ── CLI ─────────────────────────────────────────────────────────────

interface ParsedArgs {
  action?: string;
  params?: string;
  help?: boolean;
}

function parseArgs(argv: readonly string[]): ParsedArgs {
  const out: ParsedArgs = {};
  let i = 0;
  while (i < argv.length) {
    const a = argv[i];
    if (a === undefined) {
      i++;
      continue;
    }
    if (a === "-h" || a === "--help") {
      out.help = true;
      i++;
      continue;
    }
    let name: string;
    let value: string | undefined;
    if (a.startsWith("--")) {
      const eq = a.indexOf("=");
      if (eq >= 0) {
        name = a.slice(2, eq);
        value = a.slice(eq + 1);
        i++;
      } else {
        name = a.slice(2);
        value = argv[i + 1];
        i += 2;
      }
    } else {
      throw new Error(`unrecognized argument: ${a}`);
    }
    switch (name) {
      case "action":
        out.action = value ?? "";
        break;
      case "params":
        out.params = value ?? "";
        break;
      default:
        throw new Error(`unrecognized argument: --${name}`);
    }
  }
  return out;
}

const HELP_TEXT = `usage: gcp_query.js [-h] --action {describe_db,list_services,list_topics,query_logs} [--params PARAMS]

Query GCP resources

options:
  -h, --help            show this help message and exit
  --action {describe_db,list_services,list_topics,query_logs}
                        Action to perform
  --params PARAMS       JSON string of action parameters
`;

export function main(argv: readonly string[] = process.argv.slice(2)): number {
  let args: ParsedArgs;
  try {
    args = parseArgs(argv);
  } catch (e) {
    process.stderr.write(`${(e as Error).message}\n`);
    return 2;
  }

  if (args.help) {
    process.stdout.write(HELP_TEXT);
    return 0;
  }

  if (!args.action) {
    process.stderr.write("the following arguments are required: --action\n");
    return 2;
  }

  if (!VALID_ACTIONS.has(args.action)) {
    const sorted = [...VALID_ACTIONS].sort();
    process.stderr.write(
      `argument --action: invalid choice: '${args.action}' (choose from ${sorted.map((s) => `'${s}'`).join(", ")})\n`,
    );
    return 2;
  }

  const paramsRaw = args.params ?? "{}";
  let params: Params;
  try {
    const parsed: unknown = JSON.parse(paramsRaw);
    if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) {
      throw new Error("params must be a JSON object");
    }
    params = parsed as Params;
  } catch (e) {
    const result: FetchResult = {
      status: "error",
      error_code: "VALIDATION_ERROR",
      message: `Invalid JSON in --params: ${(e as Error).message}`,
    };
    process.stdout.write(pythonJsonDumps(result, 2, true) + "\n");
    return 1;
  }

  const result = fetch(args.action, params);
  process.stdout.write(pythonJsonDumps(result, 2, true) + "\n");

  if (result["status"] !== "success") {
    return 1;
  }
  return 0;
}

const thisFile = fileURLToPath(import.meta.url);
if (process.argv[1] && path.resolve(process.argv[1]) === thisFile) {
  process.exit(main());
}
